sw/device/examples/sram_program/sram_program.c

```diff
@@ -0,0 +1,42 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdint.h>
+
+#include "sw/device/lib/base/macros.h"
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/lib/dif/dif_uart.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/runtime/print.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+static dif_uart_t uart0;
+
+enum {
+  kSramStart = TOP_EARLGREY_SRAM_CTRL_MAIN_RAM_BASE_ADDR,
+  kSramEnd = TOP_EARLGREY_SRAM_CTRL_MAIN_RAM_BASE_ADDR +
+             TOP_EARLGREY_SRAM_CTRL_MAIN_RAM_SIZE_BYTES,
+};
+
+void sram_main() {
+  // Initialize UART.
+  CHECK_DIF_OK(dif_uart_init(
+      mmio_region_from_addr(TOP_EARLGREY_UART0_BASE_ADDR), &uart0));
+  CHECK_DIF_OK(
+      dif_uart_configure(&uart0, (dif_uart_config_t){
+                                     .baudrate = kUartBaudrate,
+                                     .clk_freq_hz = kClockFreqPeripheralHz,
+                                     .parity_enable = kDifToggleDisabled,
+                                     .parity = kDifUartParityEven,
+                                 }));
+  base_uart_stdout(&uart0);
+
+  // Read the program counter and check that we are executing from SRAM.
+  uint32_t pc = 0;
+  asm volatile("auipc %[pc], 0;" : [pc] "=r"(pc));
```

Comment:
Nit: IIRC this doesn't need to be volatile. You're using it to get the PC, not for any side effects. If you didn't actually need the PC (e.g. dead store into the `pc` var) it would be fair game to optimize this asm away. Of course in this case it won't be optimized out, because you're actually using it.

Reply:
Thanks, done.